File: frontend/src/api/fieldOperations.ts
import { apiFetch } from "./client";

export type SelectRecord = { id: string; name: string; [key: string]: unknown };
export type Employee = {
  id: string;
  first_name: string;
  last_name: string;
  email: string;
  role: string | null;
  phone: string | null;
  address: string | null;
  emergency_contact_name: string | null;
  emergency_contact_phone: string | null;
  emergency_contact_relationship: string | null;
  hire_date: string | null;
  portal_role: "employee" | "operator" | "foreman" | "management";
  notes: string | null;
  status: string;
};
export type Vehicle = SelectRecord & {
  unit_number: string;
  assigned_driver_name: string | null;
  make: string | null;
  model: string | null;
  current_km: number;
  next_service_km: number | null;
  next_service_date: string | null;
  status: string;
  service_status: "current" | "due_soon" | "overdue";
};
export type FieldRecord = {
  id: string;
  record_type: string;
  project_id: string | null;
  employee_id: string | null;
  equipment_id: string | null;
  supplier_id: string | null;
  cost_code: string | null;
  work_date: string;
  title: string;
  status: string;
  severity: "none" | "low" | "medium" | "high" | "critical";
  details: Record<string, unknown>;
  document_ids: string[];
  signatures: Array<Record<string, string>>;
  alert_recipients: string[];
  submitted_by: string | null;
};
export type FieldOperationsBootstrap = {
  employees: Employee[];
  projects: SelectRecord[];
  suppliers: SelectRecord[];
  equipment: SelectRecord[];
  cost_codes: Array<{ code: string; name: string }>;
  job_workbooks: Array<{ id: string; project_id: string; status: string; created_at: string; line_count: number }>;
  production_items: Array<{
    line_key: string;
    workbook_id: string;
    project_id: string;
    cost_code: string | null;
    description: string;
    unit: string;
    estimated_quantity: number;
    installed_quantity: number;
    remaining_quantity: number;
    percent_complete: number;
    materials: Array<Record<string, unknown>>;
  }>;
  material_types: Array<{ code: string; name: string }>;
  material_movement_summary: Array<{
    project_id: string | null;
    direction: "imported" | "exported";
    material_code: string;
    material_type: string;
    loads: number;
    total_tonnes: number;
  }>;
  milestone_catalog: Array<{
    id: string;
    track: "civil" | "operator";
    name: string;
    minimum_months: number;
    practical: string[];
    written_questions: Array<{ id: string; prompt: string; options: string[] }>;
  }>;
  milestone_recognitions: Array<{
    record_id: string;
    employee_id: string;
    employee_name: string;
    milestone_name: string;
    approved_at: string;
    reward_type: string;
    reward_description: string | null;
  }>;
  paperwork_recognitions: Array<{ employee_id: string; employee_name: string; on_time_days: number }>;
  vehicles: Vehicle[];
  vehicle_logs: Array<Record<string, unknown>>;
  time_entries: Array<Record<string, unknown>>;
  records: FieldRecord[];
  certifications: Array<Record<string, unknown>>;
  alerts: Array<{ type: string; severity: string; title: string; recipients: string[] }>;
  toolbox_talk: {
    week_of: string;
    title: string;
    summary: string;
    discussion_points: string[];
    source_name: string;
    source_url: string;
  };
};

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL ?? "http://localhost:8000/api/v1";

async function request<T>(path: string, options?: RequestInit): Promise<T> {
  const response = await apiFetch(API_BASE_URL + path, {
    headers: { "Content-Type": "application/json", ...options?.headers },
    ...options,
  });
  if (!response.ok) {
    const body = await response.json().catch(() => null) as { detail?: string } | null;
    throw new Error(body?.detail ?? "Request failed with " + response.status);
  }
  return response.json() as Promise<T>;
}

export const fieldOperationsApi = {
  bootstrap: () => request<FieldOperationsBootstrap>("/field-operations/bootstrap"),
  createVehicleLog: (payload: Record<string, unknown>) =>
    request("/field-operations/vehicle-logs", { method: "POST", body: JSON.stringify(payload) }),
  updateVehicle: (id: string, payload: Record<string, unknown>) =>
    request("/field-operations/vehicles/" + id, { method: "PATCH", body: JSON.stringify(payload) }),
  createTimeEntry: (payload: Record<string, unknown>) =>
    request("/field-operations/time-entries", { method: "POST", body: JSON.stringify(payload) }),
  createRecord: (payload: Record<string, unknown>) =>
    request<FieldRecord>("/field-operations/records", { method: "POST", body: JSON.stringify(payload) }),
  signRecord: (id: string, payload: Record<string, unknown>) =>
    request<FieldRecord>("/field-operations/records/" + id + "/sign", { method: "POST", body: JSON.stringify(payload) }),
  decideMilestone: (id: string, payload: Record<string, unknown>) =>
    request<FieldRecord>("/field-operations/records/" + id + "/milestone-decision", { method: "POST", body: JSON.stringify(payload) }),
  createEmployee: (payload: Record<string, unknown>) =>
    request("/field-operations/employees", { method: "POST", body: JSON.stringify(payload) }),
  createCertification: (payload: Record<string, unknown>) =>
    request("/field-operations/certifications", { method: "POST", body: JSON.stringify(payload) }),
};

File: backend/app/models/field_operations.py
from datetime import date, datetime
from uuid import UUID

from sqlalchemy import Date, DateTime, Float, ForeignKey, Numeric, String, Text
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base
from app.db.types import JSONType
from app.models.mixins import TimestampMixin, UUIDPrimaryKeyMixin


class Vehicle(UUIDPrimaryKeyMixin, TimestampMixin, Base):
    __tablename__ = "vehicles"

    unit_number: Mapped[str] = mapped_column(String(40), unique=True, nullable=False, index=True)
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    assigned_driver_name: Mapped[str | None] = mapped_column(String(255))
    assigned_employee_id: Mapped[UUID | None] = mapped_column(ForeignKey("employees.id"))
    make: Mapped[str | None] = mapped_column(String(120))
    model: Mapped[str | None] = mapped_column(String(120))
    year: Mapped[int | None]
    vin: Mapped[str | None] = mapped_column(String(40), unique=True)
    licence_plate: Mapped[str | None] = mapped_column(String(40))
    current_km: Mapped[float] = mapped_column(Float, default=0, nullable=False)
    next_service_km: Mapped[float | None] = mapped_column(Float)
    next_service_date: Mapped[date | None] = mapped_column(Date)
    status: Mapped[str] = mapped_column(String(40), default="active", nullable=False)
    notes: Mapped[str | None] = mapped_column(Text)


class VehicleLog(UUIDPrimaryKeyMixin, TimestampMixin, Base):
    __tablename__ = "vehicle_logs"

    vehicle_id: Mapped[UUID] = mapped_column(ForeignKey("vehicles.id"), nullable=False, index=True)
    employee_id: Mapped[UUID | None] = mapped_column(ForeignKey("employees.id"))
    project_id: Mapped[UUID | None] = mapped_column(ForeignKey("projects.id"))
    log_type: Mapped[str] = mapped_column(String(40), nullable=False, index=True)
    entry_date: Mapped[date] = mapped_column(Date, nullable=False, index=True)
    odometer_km: Mapped[float | None] = mapped_column(Float)
    litres: Mapped[float | None] = mapped_column(Float)
    amount: Mapped[float | None] = mapped_column(Numeric(12, 2))
    vendor: Mapped[str | None] = mapped_column(String(255))
    details: Mapped[str | None] = mapped_column(Text)
    document_ids: Mapped[list] = mapped_column(JSONType, default=list, nullable=False)


class EmployeeCertification(UUIDPrimaryKeyMixin, TimestampMixin, Base):
    __tablename__ = "employee_certifications"

    employee_id: Mapped[UUID] = mapped_column(ForeignKey("employees.id"), nullable=False, index=True)
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    issuer: Mapped[str | None] = mapped_column(String(255))
    certificate_number: Mapped[str | None] = mapped_column(String(120))
    issued_date: Mapped[date | None] = mapped_column(Date)
    expiry_date: Mapped[date | None] = mapped_column(Date, index=True)
    document_id: Mapped[UUID | None] = mapped_column(ForeignKey("documents.id"))
    notes: Mapped[str | None] = mapped_column(Text)


class TimeEntry(UUIDPrimaryKeyMixin, TimestampMixin, Base):
    __tablename__ = "time_entries"

    employee_id: Mapped[UUID] = mapped_column(ForeignKey("employees.id"), nullable=False, index=True)
    project_id: Mapped[UUID] = mapped_column(ForeignKey("projects.id"), nullable=False, index=True)
    cost_code: Mapped[str] = mapped_column(String(32), nullable=False, index=True)
    work_date: Mapped[date] = mapped_column(Date, nullable=False, index=True)
    regular_hours: Mapped[float] = mapped_column(Float, default=0, nullable=False)
    overtime_hours: Mapped[float] = mapped_column(Float, default=0, nullable=False)
    entry_type: Mapped[str] = mapped_column(String(40), default="employee", nullable=False)
    notes: Mapped[str | None] = mapped_column(Text)
    status: Mapped[str] = mapped_column(String(40), default="submitted", nullable=False)
    submitted_by: Mapped[str | None] = mapped_column(String(255))


class FieldRecord(UUIDPrimaryKeyMixin, TimestampMixin, Base):
    __tablename__ = "field_records"

    record_type: Mapped[str] = mapped_column(String(80), nullable=False, index=True)
    project_id: Mapped[UUID | None] = mapped_column(ForeignKey("projects.id"), index=True)
    employee_id: Mapped[UUID | None] = mapped_column(ForeignKey("employees.id"), index=True)
    equipment_id: Mapped[UUID | None] = mapped_column(ForeignKey("equipment.id"))
    supplier_id: Mapped[UUID | None] = mapped_column(ForeignKey("suppliers.id"))
    cost_code: Mapped[str | None] = mapped_column(String(32), index=True)
    work_date: Mapped[date] = mapped_column(Date, nullable=False, index=True)
    title: Mapped[str] = mapped_column(String(255), nullable=False)
    status: Mapped[str] = mapped_column(String(40), default="submitted", nullable=False)
    severity: Mapped[str] = mapped_column(String(40), default="none", nullable=False)
    details: Mapped[dict] = mapped_column(JSONType, default=dict, nullable=False)
    document_ids: Mapped[list] = mapped_column(JSONType, default=list, nullable=False)
    signatures: Mapped[list] = mapped_column(JSONType, default=list, nullable=False)
    alert_recipients: Mapped[list] = mapped_column(JSONType, default=list, nullable=False)
    submitted_by: Mapped[str | None] = mapped_column(String(255))
    resolved_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
